File: backend/src/routes/meals.js
import express from "express";
import { MEAL_TYPES } from "../constants/mealTypes.js";
import { ROLES } from "../constants/roles.js";
import { requireAuth, requireRole } from "../middleware/auth.js";
import { getUserById } from "../services/userService.js";
import { optOut, optIn, getUserMealStatus } from "../services/mealService.js";
import { getAvailableMeals } from "../services/availabilityService.js";
import { enforceCutoff } from "../middleware/cutoff.js";
import { broadcast } from "../services/sseService.js";
import { logAction } from "../services/auditService.js";
import { getSettings } from "../services/settingsService.js";

const router = express.Router();

function todayString() {
    const now = new Date();
    const y = now.getFullYear();
    const m = String(now.getMonth() + 1).padStart(2, "0");
    const d = String(now.getDate()).padStart(2, "0");
    return `${y}-${m}-${d}`;
}

const mealDateExtractor = (req) => {
    const date = req.body.date;
    return (date && date.trim()) ? date : todayString();
};

function isDateBeyondForwardWindow(dateStr, maxDays) {
    const today = new Date();
    today.setHours(0, 0, 0, 0);
    const target = new Date(dateStr + "T00:00:00");
    const diffDays = Math.round((target - today) / (1000 * 60 * 60 * 24));
    return diffDays > maxDays;
}

/**
 * GET /api/meals?date=YYYY-MM-DD
 * Returns available meals for the date and the logged-in user's status for each.
 */
router.get("/", requireAuth, (req, res) => {
    const date = req.query.date;
    if (!date) {
        return res.status(400).json({ error: "date query param is required" });
    }

    const user = req.session.user;
    const availableMeals = getAvailableMeals(date);
    const userStatus = getUserMealStatus(user.id, availableMeals, date);

    const meals = availableMeals.map((m) => ({
        type: m.type,
        default: m.default,
        status: userStatus[m.type],
    }));

    res.json({ date, meals });
});

/**
 * POST /api/meals/:mealType/opt-out
 * Employee opts out of a meal. Body: { date? }
 */
router.post("/:mealType/opt-out", requireAuth, enforceCutoff(mealDateExtractor), (req, res) => {
    const { mealType } = req.params;
    const user = req.session.user;
    const date = req.body.date;
    const targetDate = date || todayString();

    if (!MEAL_TYPES.includes(mealType)) {
        return res.status(400).json({ error: `Invalid meal type: ${mealType}` });
    }

    if (user.role === ROLES.EMPLOYEE) {
        const { maxForwardPlanningDays } = getSettings();
        if (isDateBeyondForwardWindow(targetDate, maxForwardPlanningDays)) {
            return res.status(400).json({ error: "Date is beyond the allowed forward planning window." });
        }
    }

    const available = getAvailableMeals(targetDate);
    if (available.length === 0) {
        return res.status(400).json({ error: "No meals available on this date (holiday or office closed)" });
    }
    if (!available.some(m => m.type === mealType)) {
        return res.status(400).json({ error: `${mealType} is not available on ${targetDate}` });
    }

    optOut(user.id, mealType, user.id, date);
    broadcast("headcount-update", { date: targetDate });

    try {
        logAction({ actorId: user.id, actorName: user.name, targetUserId: user.id, actionType: "MEAL_OPT_OUT", details: { date: targetDate, mealType } });
    } catch (e) {
        console.error("Audit write failed:", e);
    }

    res.json({ message: `Opted out of ${mealType}` });
});

/**
 * POST /api/meals/:mealType/opt-in
 * Employee opts back in to a meal. Body: { date? }
 */
router.post("/:mealType/opt-in", requireAuth, enforceCutoff(mealDateExtractor), (req, res) => {
    const { mealType } = req.params;
    const user = req.session.user;
    const date = req.body.date;
    const targetDate = date || todayString();

    if (!MEAL_TYPES.includes(mealType)) {
        return res.status(400).json({ error: `Invalid meal type: ${mealType}` });
    }

    if (user.role === ROLES.EMPLOYEE) {
        const { maxForwardPlanningDays } = getSettings();
        if (isDateBeyondForwardWindow(targetDate, maxForwardPlanningDays)) {
            return res.status(400).json({ error: "Date is beyond the allowed forward planning window." });
        }
    }

    const available = getAvailableMeals(targetDate);
    if (available.length === 0) {
        return res.status(400).json({ error: "No meals available on this date (holiday or office closed)" });
    }
    if (!available.some(m => m.type === mealType)) {
        return res.status(400).json({ error: `${mealType} is not available on ${targetDate}` });
    }

    optIn(user.id, mealType, user.id, date);
    broadcast("headcount-update", { date: targetDate });

    try {
        logAction({ actorId: user.id, actorName: user.name, targetUserId: user.id, actionType: "MEAL_OPT_IN", details: { date: targetDate, mealType } });
    } catch (e) {
        console.error("Audit write failed:", e);
    }

    res.json({ message: `Opted in to ${mealType}` });
});

/**
 * POST /api/meals/override
 * TL (own team) or Admin (any user) overrides a single meal.
 * Body: { targetUserId, mealType, status, date? }
 */
router.post(
    "/override",
    requireAuth,
    requireRole([ROLES.TEAM_LEAD, ROLES.ADMIN]),
    (req, res) => {
        const { targetUserId, mealType, status, date } = req.body;
        const currentUser = req.session.user;

        if (!targetUserId || !mealType || !["IN", "OUT"].includes(status)) {
            return res.status(400).json({
                error: "targetUserId, mealType, and status (IN/OUT) are required",
            });
        }

        if (!MEAL_TYPES.includes(mealType)) {
            return res.status(400).json({ error: `Invalid meal type: ${mealType}` });
        }

        const targetUser = getUserById(targetUserId);
        if (!targetUser) {
            return res.status(404).json({ error: "Target user not found" });
        }

        if (
            currentUser.role === ROLES.TEAM_LEAD &&
            currentUser.teamId !== targetUser.teamId
        ) {
            return res.status(403).json({
                error: "Team Leads can only override their own team members",
            });
        }

        if (status === "OUT") {
            optOut(targetUserId, mealType, currentUser.id, date);
        } else {
            optIn(targetUserId, mealType, currentUser.id, date);
        }

        const targetDate = date || todayString();
        broadcast("headcount-update", { date: targetDate });

        try {
            logAction({ actorId: currentUser.id, actorName: currentUser.name, targetUserId, actionType: "MEAL_OVERRIDE", details: { date: targetDate, mealType, status } });
        } catch (e) {
            console.error("Audit write failed:", e);
        }

        res.json({ message: `Overrode ${targetUser.name}'s ${mealType} to ${status}` });
    }
);

/**
 * POST /api/meals/bulk-override
 * Bulk override meal participation for multiple users, meals, and dates.
 * Body: { userIds, mealTypes, status, startDate, endDate }
 * Rejects entirely if any target user is outside actor's scope.
 */
router.post(
    "/bulk-override",
    requireAuth,
    requireRole([ROLES.TEAM_LEAD, ROLES.ADMIN]),
    (req, res) => {
        const { userIds, mealTypes, status, startDate, endDate } = req.body;
        const currentUser = req.session.user;

        if (!userIds || !mealTypes || !status || !startDate || !endDate) {
            return res.status(400).json({
                error: "userIds, mealTypes, status, startDate, and endDate are required",
            });
        }

        if (!Array.isArray(userIds) || !Array.isArray(mealTypes)) {
            return res.status(400).json({ error: "userIds and mealTypes must be arrays" });
        }

        if (!["IN", "OUT"].includes(status)) {
            return res.status(400).json({ error: "status must be IN or OUT" });
        }

        if (startDate > endDate) {
            return res.status(400).json({ error: "startDate must be before or equal to endDate" });
        }

        for (const mt of mealTypes) {
            if (!MEAL_TYPES.includes(mt)) {
                return res.status(400).json({ error: `Invalid meal type: ${mt}` });
            }
        }

        const targetUsers = [];
        for (const uid of userIds) {
            const user = getUserById(uid);
            if (!user) {
                return res.status(404).json({ error: `User not found: ${uid}` });
            }
            targetUsers.push(user);
        }

        if (currentUser.role === ROLES.TEAM_LEAD) {
            const outOfScope = targetUsers.find((u) => u.teamId !== currentUser.teamId);
            if (outOfScope) {
                return res.status(403).json({
                    error: "Team Leads can only bulk override their own team members",
                });
            }
        }

        const dates = [];
        let current = startDate;
        while (current <= endDate) {
            dates.push(current);
            const [y, m, d] = current.split("-").map(Number);
            current = new Date(Date.UTC(y, m - 1, d + 1)).toISOString().split("T")[0];
        }

        const action = status === "OUT" ? optOut : optIn;
        let count = 0;
        for (const date of dates) {
            for (const uid of userIds) {
                for (const mt of mealTypes) {
                    action(uid, mt, currentUser.id, date);
                    count++;
                }
            }
        }

        for (const date of dates) {
            broadcast("headcount-update", { date });
        }

        // One audit entry for the whole batch
        try {
            logAction({ actorId: currentUser.id, actorName: currentUser.name, targetUserId: null, actionType: "BULK_OVERRIDE", details: { startDate, endDate, mealTypes, status, userIds } });
        } catch (e) {
            console.error("Audit write failed:", e);
        }

        res.json({ message: `Applied ${count} overrides`, count });
    }
);

export default router;

File: mhp-python/app/routes/meals.py
from datetime import date as DateType, datetime, timezone
from fastapi import APIRouter, HTTPException, Depends, status

from app.auth import get_current_user, require_roles
from app.constants import Roles, MealTypes
from app.models import OptMealRequest, MealOverrideRequest, BulkOverrideRequest
from app.services.availability_service import get_available_meals
from app.services.meal_service import opt_in, opt_out, get_user_meal_status
from app.services.user_service import get_user_by_id
from app.services.settings_service import get_settings
from app.services.audit_service import log_action

router = APIRouter()


def today() -> str:
    return DateType.today().isoformat()


def _check_cutoff(target_date: str, user_role: str):
    """
    Block EMPLOYEE and LOGISTICS after 21:00 the day before target date.
    TL and ADMIN bypass this entirely.
    Cutoff is hardcoded to 21:00 in the Python version (not read from
    settings) to avoid an extra DB read on every mutation.
    """
    if user_role in (Roles.TEAM_LEAD, Roles.ADMIN):
        return
    y, m, d = map(int, target_date.split("-"))
    cutoff = datetime(y, m, d - 1, 21, 0, 0, tzinfo=timezone.utc)
    if datetime.now(timezone.utc) > cutoff:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail=f"Changes for {target_date} are locked (cutoff was 21:00 the day before).",
        )


def _check_forward_window(target_date: str, user_role: str):
    """EMPLOYEE cannot plan beyond maxForwardPlanningDays days ahead."""
    if user_role != Roles.EMPLOYEE:
        return
    max_days = get_settings().get("maxForwardPlanningDays", 14)
    delta = (DateType.fromisoformat(target_date) - DateType.today()).days
    if delta > max_days:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Date is beyond the allowed forward planning window.",
        )


def _check_past(target_date: str):
    if target_date < today():
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Cannot modify records for past dates.")


@router.get("")
def get_meals(date: str, user: dict = Depends(get_current_user)):
    available = get_available_meals(date)
    user_status = get_user_meal_status(user["id"], available, date)
    meals = [{"type": m["type"], "default": m["default"], "status": user_status[m["type"]]} for m in available]
    return {"date": date, "meals": meals}


@router.post("/{meal_type}/opt-out")
def meal_opt_out(meal_type: str, body: OptMealRequest, user: dict = Depends(get_current_user)):
    target_date = body.date or today()
    if meal_type not in MealTypes.ALL:
        raise HTTPException(status_code=400, detail=f"Invalid meal type: {meal_type}")
    _check_past(target_date)
    _check_cutoff(target_date, user["role"])
    _check_forward_window(target_date, user["role"])

    available = get_available_meals(target_date)
    if not available:
        raise HTTPException(status_code=400, detail="No meals available on this date.")
    if not any(m["type"] == meal_type for m in available):
        raise HTTPException(status_code=400, detail=f"{meal_type} is not available on {target_date}.")

    opt_out(user["id"], meal_type, user["id"], target_date)
    try:
        log_action(user["id"], user["name"], user["id"], "MEAL_OPT_OUT", {"date": target_date, "mealType": meal_type})
    except Exception:
        pass
    return {"message": f"Opted out of {meal_type}"}


@router.post("/{meal_type}/opt-in")
def meal_opt_in(meal_type: str, body: OptMealRequest, user: dict = Depends(get_current_user)):
    target_date = body.date or today()
    if meal_type not in MealTypes.ALL:
        raise HTTPException(status_code=400, detail=f"Invalid meal type: {meal_type}")
    _check_past(target_date)
    _check_cutoff(target_date, user["role"])
    _check_forward_window(target_date, user["role"])

    available = get_available_meals(target_date)
    if not available:
        raise HTTPException(status_code=400, detail="No meals available on this date.")
    if not any(m["type"] == meal_type for m in available):
        raise HTTPException(status_code=400, detail=f"{meal_type} is not available on {target_date}.")

    opt_in(user["id"], meal_type, user["id"], target_date)
    try:
        log_action(user["id"], user["name"], user["id"], "MEAL_OPT_IN", {"date": target_date, "mealType": meal_type})
    except Exception:
        pass
    return {"message": f"Opted in to {meal_type}"}


@router.post("/override")
def meal_override(
    body: MealOverrideRequest,
    user: dict = Depends(require_roles([Roles.TEAM_LEAD, Roles.ADMIN])),
):
    target_date = body.date or today()
    if body.mealType not in MealTypes.ALL:
        raise HTTPException(status_code=400, detail=f"Invalid meal type: {body.mealType}")
    if body.status not in ("IN", "OUT"):
        raise HTTPException(status_code=400, detail="status must be IN or OUT")

    target_user = get_user_by_id(body.targetUserId)
    if not target_user:
        raise HTTPException(status_code=404, detail="Target user not found")

    # TL can only override their own team
    if user["role"] == Roles.TEAM_LEAD and user["teamId"] != target_user.get("teamId"):
        raise HTTPException(status_code=403, detail="Team Leads can only override their own team members")

    if body.status == "OUT":
        opt_out(body.targetUserId, body.mealType, user["id"], target_date)
    else:
        opt_in(body.targetUserId, body.mealType, user["id"], target_date)

    try:
        log_action(user["id"], user["name"], body.targetUserId, "MEAL_OVERRIDE",
                   {"date": target_date, "mealType": body.mealType, "status": body.status})
    except Exception:
        pass
    return {"message": f"Overrode {target_user['name']}'s {body.mealType} to {body.status}"}


@router.post("/bulk-override")
def bulk_override(
    body: BulkOverrideRequest,
    user: dict = Depends(require_roles([Roles.TEAM_LEAD, Roles.ADMIN])),
):
    if body.status not in ("IN", "OUT"):
        raise HTTPException(status_code=400, detail="status must be IN or OUT")
    if body.startDate > body.endDate:
        raise HTTPException(status_code=400, detail="startDate must be <= endDate")
    for mt in body.mealTypes:
        if mt not in MealTypes.ALL:
            raise HTTPException(status_code=400, detail=f"Invalid meal type: {mt}")

    target_users = []
    for uid in body.userIds:
        u = get_user_by_id(uid)
        if not u:
            raise HTTPException(status_code=404, detail=f"User not found: {uid}")
        target_users.append(u)

    if user["role"] == Roles.TEAM_LEAD:
        out_of_scope = next((u for u in target_users if u.get("teamId") != user["teamId"]), None)
        if out_of_scope:
            raise HTTPException(status_code=403, detail="Team Leads can only bulk override their own team members")

    # Build date range
    dates = []
    current = DateType.fromisoformat(body.startDate)
    end = DateType.fromisoformat(body.endDate)
    while current <= end:
        dates.append(current.isoformat())
        from datetime import timedelta
        current += timedelta(days=1)

    action = opt_out if body.status == "OUT" else opt_in
    count = 0
    for d in dates:
        for uid in body.userIds:
            for mt in body.mealTypes:
                action(uid, mt, user["id"], d)
                count += 1

    try:
        log_action(user["id"], user["name"], None, "BULK_OVERRIDE",
                   {"startDate": body.startDate, "endDate": body.endDate,
                    "mealTypes": body.mealTypes, "status": body.status, "userIds": body.userIds})
    except Exception:
        pass
    return {"message": f"Applied {count} overrides", "count": count}
